lib/Runtime/Library/ForInObjectEnumerator.cpp

```diff
@@ -182,7 +207,32 @@ namespace Js
                 if (TestAndSetEnumerated(propertyId) //checks if the property is already enumerated or not
                     && (attributes & PropertyEnumerable))
                 {
-                    return currentIndex;
+                    bool propertyShadowed = false;
```

Comment:
My suggestion is that we still skip to the first prototype with enumerable properties, but don't skip non-visited prototype for downward lookup (the assumption is even some JS framework may add method to built-in prototype, like `Object.prototype`, the number is very small so the downward lookup number is also small, not like the test case you've made before). The revision is excluding prototype from downward lookup when it is visited (this led to 3.5 times slow in your previous example), it eliminates the unnecessary downward lookup if the most of the prototype have enumerable properties (they will be visited in sequential).

In some extreme cases like a lot of properties needs downward lookup (like if any code only adds tons of methods to `Object.prototype`, we could add a heuristic that fallback to full lookup, as you suggested, after some number of downward search.

This is trying avoid full lookup if possible, but I think the plain heuristic for full lookup (do full search if there is any enumerable properties in the chain) should be good in practical.